### PR TITLE
messaging, papercuts, prevent & heal from invalid states

### LINKED ISSUE #48: when loading, validate that the state is correct for the steps
each operation should have a `validate` function that checks that the `data` and the user's machine has the state it would expect given the current `steps`. for example, `runScript.validate` would check that a cloned repository with a correctly named branch exists in the workspace folder.

if an operation doesn't validate, it needs to be re-run on load.

### LINKED ISSUE #50: improve messaging and help text
nori doesn't exactly hold your hand right now. it needs some explanatory text for what each stage (loading/creating state file, running operations) means. we should think about UX for first use: currently it just asks you `name?` which isn't helpful

### LINKED ISSUE #59: better behaviour if local clone/branch exists
currently `run-script` errors if the branch already exists, and if the clone already exists it only checks out master before creating the new branch. it should hard reset, pull from origin, & delete existing branches. goes hand in hand with #56 and #48

### LINKED ISSUE #65: Status 422 when the pull request already exists
```
{ HttpError: Validation Failed
    at response.text.then.message (/Users/samuel.parkinson/Repositories/Financial-Times/transformation-runner/node_modules/@octokit/request/lib/request.js:56:27)
    at processTicksAndRejections (internal/process/task_queues.js:86:5)
  name: 'HttpError',
  status: 422,
  headers:
   { 'access-control-allow-origin': '*',
     'access-control-expose-headers':
      'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type',
     connection: 'close',
     'content-length': '184',
     'content-security-policy': "default-src 'none'",
     'content-type': 'application/json; charset=utf-8',
     date: 'Thu, 11 Apr 2019 15:46:15 GMT',
     'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin',
     server: 'GitHub.com',
     status: '422 Unprocessable Entity',
     'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
     'x-accepted-oauth-scopes': '',
     'x-content-type-options': 'nosniff',
     'x-frame-options': 'deny',
     'x-github-media-type': 'github.inertia-preview; format=json',
     'x-github-request-id': '7C48:0D53:0E1A:1FA3:5CAF6147',
     'x-oauth-scopes': 'repo',
     'x-ratelimit-limit': '5000',
     'x-ratelimit-remaining': '4960',
     'x-ratelimit-reset': '1555001014',
     'x-xss-protection': '1; mode=block' },
  request:
   { method: 'POST',
     url:
      'https://api.github.com/repos/Financial-Times/n-content-body/pulls',
     headers:
      { accept: 'application/vnd.github.inertia-preview+json',
        'user-agent': 'octokit.js/16.19.0 Node.js/11.13.0 (macOS Mojave; x64)',
        authorization: 'token [REDACTED]',
        'content-type': 'application/json; charset=utf-8' },
     body:
      '{"head":"unpin-ngage-patch","base":"master","title":"Unpin @financial-times/n-gage","body":""}',
     request: { validate: [Object] } },
  errors:
   [ { resource: 'PullRequest', field: 'head', code: 'invalid' } ],
  documentation_url:
   'https://developer.github.com/v3/pulls/#create-a-pull-request' }
```

### LINKED ISSUE #68: Better error messages
We should improve upon error situations and provide more helpful error messages.

```
✔ path to a text file of repositories · test
{ [Error: EISDIR: illegal operation on a directory, read] errno: -21, code: 'EISDIR', syscall: 'read' }
```

### LINKED ISSUE #74: Clears state file when an operation errors
..._possibly_. seen when @jenniferemshepherd was using it. can't work out why it would

### PR DESCRIPTION
**Pull request by quarterto**
this PR is aimed at fixing common papercuts with nori, by improving the help text, adding first-run guidance, preventing common inconsistent states, and improving argument validation. 

fixes #68. fixes #50. fixes #48. fixes #65. fixes #59. fixes #74.

### CODE REVIEW

**Comment by taktran** [comment]
Could I get a description pls :)

File: src/lib/promise-all-errors.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// like Promise.all, but waits for every promise to complete even if one errors
```
Comment (JakeChampion) [inline src/lib/promise-all-errors.js]: There is a method for this that will soon be in JavaScript, it is called `Promise.allSettled`.

A polyfill for this is available and works on all NodeJS version you currently support --> https://www.npmjs.com/package/promise.allsettled

The benefit to using the polyfill would be the ability to remove it when this project moves to a version of NodeJS which supports the method natively :-)

Comment (adambraimbridge) [reply src/lib/promise-all-errors.js]: So that's that all settled then

File: src/commands/run-script.js [hunk 2]
```diff
@@ -27,31 +42,34 @@ exports.args = [
...
-	}
-
 	// must be serial until https://github.com/Financial-Times/tooling-helpers/issues/74
 	// is resolved (or, add workingDirectory to all the options of the git methods)
 	for (const repository of state.repos) {
 		const repoLabel = `${repository.owner}/${repository.name}`
 		git.defaults({ workingDirectory: repository.clone })
 
 		try {
+			// if the branch we're trying to create already exists, create `branch-1`
```
Comment (taktran) [inline src/commands/run-script.js]: 😻

Comment (jenniferemshepherd) [reply src/commands/run-script.js]: 🙌

**Pull request merged**